// @nativescript/webpack/dist/transformers/NativeClass/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const typescript_1 = __importDefault(require("typescript"));
/**
 * A TypeScript transform that compiles classes marked with @NativeClass as es5 & commonjs
 *
 * @param ctx
 */
function default_1(ctx) {
    function isNativeClassExtension(node) {
        return (node.decorators &&
            node.decorators.filter((d) => {
                const fullText = d.getFullText().trim();
                return fullText.indexOf('@NativeClass') > -1;
            }).length > 0);
    }
    function visitNode(node) {
        if (typescript_1.default.isClassDeclaration(node) && isNativeClassExtension(node)) {
            return createHelper(node);
        }
        return typescript_1.default.visitEachChild(node, visitNode, ctx);
    }
    function createHelper(node) {
        // we remove the decorator for now!
        return typescript_1.default.createIdentifier(typescript_1.default
            .transpileModule(node.getText().replace(/@NativeClass(\((.|\n)*?\))?/gm, ''), {
            compilerOptions: {
                noEmitHelpers: true,
                module: typescript_1.default.ModuleKind.CommonJS,
                target: typescript_1.default.ScriptTarget.ES5,
            },
        })
            .outputText.replace(/(Object\.defineProperty\(.*?{.*?)(enumerable:\s*false)(.*?}\))/gs, '$1enumerable: true$3'));
    }
    return (source) => typescript_1.default.updateSourceFileNode(source, typescript_1.default.visitNodes(source.statements, visitNode));
}
exports.default = default_1;
