// @nativescript/angular/esm2020/lib/cdk/frame-page/page.directive.mjs
import { Directive, ElementRef } from '@angular/core';
import { Page } from '@nativescript/core';
import { PageService } from './page.service';
import * as i0 from "@angular/core";
export function customPageFactory(v) {
    return v.element.nativeElement;
}
export class PageDirective {
    constructor(element) {
        this.element = element;
    }
}
PageDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.2.3", ngImport: i0, type: PageDirective, deps: [{ token: i0.ElementRef }], target: i0.ɵɵFactoryTarget.Directive });
PageDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "14.2.3", type: PageDirective, selector: "Page", providers: [
        {
            provide: Page,
            useFactory: customPageFactory,
            deps: [PageDirective],
        },
        PageService,
    ], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.2.3", ngImport: i0, type: PageDirective, decorators: [{
            type: Directive,
            args: [{
                    // eslint-disable-next-line @angular-eslint/directive-selector
                    selector: 'Page',
                    providers: [
                        {
                            provide: Page,
                            useFactory: customPageFactory,
                            deps: [PageDirective],
                        },
                        PageService,
                    ],
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef }]; } });
