// @nativescript/angular/esm2020/lib/http/ns-http-backend.mjs
import { Injectable } from '@angular/core';
import { HttpResponse, HttpErrorResponse, HttpXhrBackend } from '@angular/common/http';
import { XhrFactory } from '@angular/common';
import { NSFileSystem } from '../file-system/ns-file-system';
import { isLocalRequest, processLocalFileRequest } from './http-utils';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
import * as i2 from "../file-system/ns-file-system";
export class NsHttpBackEnd extends HttpXhrBackend {
    constructor(xhrFactory, nsFileSystem) {
        super(xhrFactory);
        this.nsFileSystem = nsFileSystem;
    }
    handle(req) {
        let result;
        if (isLocalRequest(req.url)) {
            result = this.handleLocalFileRequest(req.url);
        }
        else {
            result = super.handle(req);
        }
        return result;
    }
    handleLocalFileRequest(url) {
        return processLocalFileRequest(url, this.nsFileSystem, createSuccessResponse, createErrorResponse);
    }
}
NsHttpBackEnd.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.2.3", ngImport: i0, type: NsHttpBackEnd, deps: [{ token: i1.XhrFactory }, { token: i2.NSFileSystem }], target: i0.ɵɵFactoryTarget.Injectable });
NsHttpBackEnd.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "14.2.3", ngImport: i0, type: NsHttpBackEnd });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.2.3", ngImport: i0, type: NsHttpBackEnd, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: i1.XhrFactory }, { type: i2.NSFileSystem }]; } });
function createSuccessResponse(url, body, status) {
    return new HttpResponse({
        url,
        body,
        status,
        statusText: 'OK',
    });
}
function createErrorResponse(url, body, status) {
    return new HttpErrorResponse({
        url,
        error: body,
        status,
        statusText: 'ERROR',
    });
}
