// @nativescript/angular/esm2020/lib/platform-ref.mjs
import { PlatformRef } from '@angular/core';
import { runNativeScriptAngularApp } from './application';
export class NativeScriptPlatformRefProxy extends PlatformRef {
    constructor(platform, launchView) {
        super();
        this.platform = platform;
        this.launchView = launchView;
    }
    bootstrapModuleFactory(moduleFactory) {
        this.options = {
            appModuleBootstrap: () => this.platform.bootstrapModuleFactory(moduleFactory),
            launchView: () => this.launchView,
        };
        runNativeScriptAngularApp(this.options);
        return null;
    }
    bootstrapModule(moduleType, compilerOptions) {
        this.options = {
            appModuleBootstrap: () => this.platform.bootstrapModule(moduleType, compilerOptions),
        };
        runNativeScriptAngularApp(this.options);
        return null;
    }
    onDestroy(callback) {
        this.platform.onDestroy(callback);
    }
    get injector() {
        return this.platform.injector;
    }
    destroy() {
        this.platform.destroy();
    }
    get destroyed() {
        return this.platform.destroyed;
    }
}
