// @nativescript/core/fps-meter/fps-native.android.js
import { Device } from '../platform';
export class FPSCallback {
    constructor(onFrame) {
        this.running = false;
        this.onFrame = onFrame;
        this.sdkVersion = parseInt(Device.sdkVersion);
        if (this.sdkVersion >= 24 && this._isNativeFramesSupported()) {
            this.impl = (nanos) => {
                this.handleFrame(nanos);
            };
        }
        else {
            this.impl = new android.view.Choreographer.FrameCallback({
                doFrame: (nanos) => {
                    this.handleFrame(nanos);
                },
            });
        }
    }
    _isNativeFramesSupported() {
        return typeof global.__postFrameCallback === 'function' && typeof global.__removeFrameCallback === 'function';
    }
    start() {
        if (this.running) {
            return;
        }
        if (this.sdkVersion >= 24 && this._isNativeFramesSupported()) {
            global.__postFrameCallback(this.impl);
        }
        else {
            android.view.Choreographer.getInstance().postFrameCallback(this.impl);
        }
        this.running = true;
    }
    stop() {
        if (!this.running) {
            return;
        }
        if (this.sdkVersion >= 24 && this._isNativeFramesSupported()) {
            global.__removeFrameCallback(this.impl);
        }
        else {
            android.view.Choreographer.getInstance().removeFrameCallback(this.impl);
        }
        this.running = false;
    }
    handleFrame(nanos) {
        if (!this.running) {
            return;
        }
        // divide by 1 000 000 since the parameter is in nanoseconds
        this.onFrame(nanos / 1000000);
        // add the FrameCallback instance again since it is automatically removed from the Choreographer
        if (this.sdkVersion >= 24 && this._isNativeFramesSupported()) {
            global.__postFrameCallback(this.impl);
        }
        else {
            android.view.Choreographer.getInstance().postFrameCallback(this.impl);
        }
    }
}
