// @nativescript/angular/esm2020/lib/cdk/dialog/dialog-module.mjs
import { NgModule } from '@angular/core';
import { NativeDialogCloseDirective } from './dialog-content-directives';
import { NativeDialogService } from './dialog-services';
import * as i0 from "@angular/core";
export class NativeDialogModule {
}
NativeDialogModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.2.3", ngImport: i0, type: NativeDialogModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NativeDialogModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.2.3", ngImport: i0, type: NativeDialogModule, declarations: [NativeDialogCloseDirective], exports: [NativeDialogCloseDirective] });
NativeDialogModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.2.3", ngImport: i0, type: NativeDialogModule, providers: [NativeDialogService] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.2.3", ngImport: i0, type: NativeDialogModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NativeDialogCloseDirective],
                    exports: [NativeDialogCloseDirective],
                    providers: [NativeDialogService],
                }]
        }] });
