// @nativescript/angular/esm2020/lib/cdk/dialog/dialog-config.mjs
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
/**
 * Configuration for opening a modal dialog with the MatDialog service.
 */
export class NativeDialogConfig {
    constructor() {
        /** Whether the dialog has a backdrop. */
        this.hasBackdrop = true;
        /** Whether the user can use escape or clicking on the backdrop to close the modal. */
        this.disableClose = false;
        /** Data being injected into the child component. */
        this.data = null;
        /**
         * Whether the dialog should close when the user goes backwards/forwards in history.
         * Note that this usually doesn't include clicking on links (unless the user is using
         * the `HashLocationStrategy`).
         */
        this.closeOnNavigation = true;
        this.nativeOptions = {};
        // TODO(jelbourn): add configuration for lifecycle hooks, ARIA labelling.
    }
}
